editors/code/src/main.ts

```diff
@@ -109,3 +99,130 @@ export async function deactivate() {
     await ctx?.client?.stop();
     ctx = undefined;
 }
+
+async function bootstrap(config: Config, state: PersistentState): Promise<string> {
+    await bootstrapExtension(config, state);
+    const path = await bootstrapServer(config, state);
+    return path;
+}
+
+async function bootstrapExtension(config: Config, state: PersistentState): Promise<void> {
+    if (config.channel === "stable") {
+        if (config.extensionReleaseTag === NIGHTLY_TAG) {
+            vscode.window.showWarningMessage(`You are running a nightly version of rust-analyzer extension.
+To switch to stable, uninstall the extension and re-install it from the marketplace`);
+        }
+        return;
+    };
+
+    const lastCheck = state.lastCheck;
+    const now = new Date().getTime();
+
+    const anHour = 60 * 60 * 1000;
+    const shouldDownloadNightly = state.releaseId === undefined || (now - (lastCheck ?? 0)) > anHour;
+
+    if (!shouldDownloadNightly) return;
+
+    const release = await fetchRelease("nightly").catch((e) => {
+        log.error(e);
+        if (state.releaseId === undefined) { // Show error only for the initial download
+            vscode.window.showErrorMessage(`Failed to download rust-analyzer nightly ${e}`);
+        }
+        return undefined;
+    });
+    if (release === undefined || release.id === state.releaseId) return;
+
+    const updateOrCancel = await vscode.window.showInformationMessage(
+        "New version of rust-analyzer (nightly) is available (requires restart).",
+        "Update", "Cancel"
+    );
+    if (updateOrCancel !== "Update") return;
+
+    const artifact = release.assets.find(artifact => artifact.name === "rust-analyzer.vsix");
+    assert(!!artifact, `Bad release: ${JSON.stringify(release)}`);
+
+    const dest = path.join(config.globalStoragePath, "rust-analyzer.vsix");
+    await download(artifact.browser_download_url, dest, "Downloading rust-analyzer extension");
+
+    await vscode.commands.executeCommand("workbench.extensions.installExtension", vscode.Uri.file(dest));
```

Comment:
Either we should use `vscodeInstallExtension()` from `util.ts` or remove that code from there.